### PR TITLE
Enable yanked dependencies with an unstable flag

### LINKED ISSUE #4225: Add flag to allow Cargo to create a lock file that depends on yanked crates
Implementation: https://github.com/rust-lang/cargo/pull/13333

---

Currently if you have a `Cargo.lock` that depends on yanked crates you are able to build the project just fine as the yanked crates still exist. However there is currently no way to create a `Cargo.lock` if the only versions of some dependency that satisfies version constraints are yanked, aside from creating a `Cargo.lock` by hand which is a really poor solution.

I propose a new flag that is added to Cargo commands capable of creating a `Cargo.lock` which will allow Cargo to add dependencies on yanked crates if it has no other option.

Inspired by https://users.rust-lang.org/t/crates-io-disable-yanking-crates-that-have-dependant-crates/11492

### PR DESCRIPTION
**Pull request by pietroalbini**
This PR adds a new unstable CLI flag, `-Z allow-yanked-deps`, that allows to depend on yanked crates even if there is no existing lockfile pointing at them (it basically disables the yanked check).

Crater needs this because it regenerates lockfiles on every run, so there are tons of crates that fail to build because they depend on yanked deps (like futures 0.2). Crater already uses unstable flags during lockfile generation, so I don't really care if this is stabilized or not.

Closes #4225
cc https://github.com/rust-lang-nursery/crater/issues/243

### CODE REVIEW

File: src/cargo/sources/registry/index.rs [hunk 1]
```diff
@@ -266,14 +266,17 @@ impl<'cfg> RegistryIndex<'cfg> {
...
         f: &mut FnMut(Summary),
     ) -> CargoResult<()> {
         let source_id = self.source_id.clone();
         let name = dep.package_name().as_str();
         let summaries = self.summaries(name, load)?;
         let summaries = summaries
             .iter()
-            .filter(|&&(_, yanked)| dep.source_id().precise().is_some() || !yanked)
+            .filter(|&&(_, yanked)| {
+                dep.source_id().precise().is_some() || allow_yanked_deps || !yanked
```
Comment (alexcrichton) [inline src/cargo/sources/registry/index.rs]: I think `self.config` is accessible here, could that be used to avoid passing in `allow_yanked_deps`?

**Comment by joshtriplett** [comment]
I don't think we *should* have a stable way to do this, and if Crater didn't need it I'd argue we shouldn't have an *unstable* way to do this either.

**Comment by Eh2406** [comment]
Several of our unstable functionality are because Crater needs it. In addition to `internal` implementation detail, and on the path to `stabilization`, we may need a `intended for rust-lang tooling` endgame for features.

**Comment by matklad** [comment]
>Several of our unstable functionality are because Crater needs it. 

I suggest delisting such features from docs and `--help` message perhaps?

File: src/cargo/sources/registry/index.rs (hunk 1 above)
Comment (pietroalbini) [reply src/cargo/sources/registry/index.rs]: Fixed.

**Comment by alexcrichton** [comment]
@pietroalbini to get a sense of scale here, how many crates fail to compile due to a dependence on yanked crates? What percentage of all crater-tested crates does this cover?

**Comment by alexcrichton** [comment]
@Eh2406 do we have other unstable functionality specifically for crater? One possible one, `-Z offline`, is something I believe we'd like to stabilize for Cargo itself at some point

**Comment by pietroalbini** [comment]
> to get a sense of scale here, how many crates fail to compile due to a dependence on yanked crates? What percentage of all crater-tested crates does this cover?

At the moment ~170, which is 1% of the crates we're testing. The other option we have right now is to manually blacklist them, or have them sit in the "Crater errors" section of the report (which is not ideal, I'd like to have nothing in it so we can catch new ones easily).

> do we have other unstable functionality specifically for crater? One possible one, `-Z offline`, is something I believe we'd like to stabilize for Cargo itself at some point

Crater doesn't use `-Z offline` because [that flag alters dependency resolution](https://github.com/rust-lang-nursery/crater/pull/178#issuecomment-359569545). Instead it uses `-Z no-index-update` (added in https://github.com/rust-lang/cargo/pull/4990), which just assumes the registry is up to date and skips updating it.

**Comment by Eh2406** [comment]
> do we have other unstable functionality specifically for crater?

when I said it it looked like https://github.com/rust-lang/cargo/pull/5961 was going to be in the list as well.

**Comment by alexcrichton** [comment]
Ah sorry I thought I replied sooner. @pietroalbini 1% seems like a pretty small number and given the historical resistance (and current resistance) to this feature, would you be ok adding such crates to the blacklist?

**Comment by pietroalbini** [comment]
I'd prefer to wait on this until I import all the GitHub repositories into Crater, to see if the numbers change.

**Comment by pietroalbini** [comment]
Ok, the number of yanked crates didn't really change. Closing this.

**Pull request closed without merging**